internal/server/data/grant.go

```diff
@@ -19,7 +40,26 @@ func CreateGrant(db GormTxn, grant *models.Grant) error {
 	case grant.Resource == "":
 		return fmt.Errorf("resource is required")
 	}
-	return add(db, grant)
+
+	// Use a savepoint so that we can query for the duplicate grant on conflict
+	if _, err := tx.Exec("SAVEPOINT beforeCreate"); err != nil {
```

Comment:
For now I'd like to only do this savepoint/rollback in the specific create functions that require it. We had added it to the shared `add` function, but we only ever saw a problem with grants. I think grants are unique here because they are immutable and have a composite unique constraint on all the interesting fields. I don't think any of our other tables have that property.

In the future we can easily move this savepoint into insert if we find a reason to.

I tried using `ON CONFLICT DO NOTHING`, but unfortunately that doesn't give any indication if the query was on conflict or if it inserted (which we need for the response code).  Maybe there's a way to do this by looking at rows modified, and then doing the select. I kept our existing approach for now to keep this change small.

Reply:
do you remove the savepoint from add somewhere?

Reply:
Nope, it's still in `add`, but `add` is no longer used with the SQL queries, so it'll be removed at the end.